planner/cascades/transformation_rules.go

```diff
@@ -589,3 +592,51 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	newJoinExpr.SetChildren(leftGroup, rightGroup)
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
+
+// MergeAdjacentProjection merge the adjacent projection.
+type MergeAdjacentProjection struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentProjection creates a new Transformation MergeAdjacentProjection.
+// The pattern of this rule is `Projection -> Projection`.
+func NewRuleMergeAdjacentProjection() Transformation {
+	rule := &MergeAdjacentProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandProjection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `proj->proj->x` to `proj->x`
+// or just keep the selection unchanged.
+func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	proj := old.GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	childGroup := old.Children[0].Group
+	child := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	if plannercore.ExprsHasSideEffects(child.Exprs) {
+		return nil, false, false, nil
+	}
+
+	replace := make(map[string]*expression.Column)
+	for i, col := range childGroup.Prop.Schema.Columns {
+		if colOrigin, ok := child.Exprs[i].(*expression.Column); ok {
+			replace[string(col.HashCode(nil))] = colOrigin
+		}
+	}
+
+	newProj := plannercore.LogicalProjection{Exprs: make([]expression.Expression, len(proj.Exprs))}.Init(proj.SCtx(), proj.SelectBlockOffset())
+	copy(newProj.Exprs, proj.Exprs)
+	newProj.SetSchema(old.GetExpr().Group.Prop.Schema.Clone())
```

Comment:
This is unnecessary, we won't use `Proj.Schema` anymore.

Reply:
Without `newProj.SetSchema`, [`*LogicalProjection.ExplainInfo()`](https://github.com/pingcap/tidb/blob/5b6422df588a345eac6cd1a72d67536c3539aa38/planner/core/explain.go#L564)  which invokes `expression.ExplainExpressionList` with `p.schema` will panic.

```go
// ExplainInfo implements Plan interface.
func (p *LogicalProjection) ExplainInfo() string {
	return expression.ExplainExpressionList(p.Exprs, p.schema)
}
```

```
PANIC: transformation_rules_test.go:142: testTransformationRuleSuite.TestProjectionElimination

... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x4031051)

/usr/local/Cellar/go/1.13/libexec/src/runtime/panic.go:679
  in gopanic
/usr/local/Cellar/go/1.13/libexec/src/runtime/panic.go:199
  in panicmem
/usr/local/Cellar/go/1.13/libexec/src/runtime/signal_unix.go:394
  in sigpanic
/Users/pingyu/workspace/go/src/github.com/pingcap/tidb/expression/explain.go:73
  in ExplainExpressionList
/Users/pingyu/workspace/go/src/github.com/pingcap/tidb/planner/core/explain.go:438
  in LogicalProjection.ExplainInfo
stringer.go:103
  in groupExprToString
stringer.go:89
  in groupToString
stringer.go:47
  in toString
stringer.go:28
  in ToString
transformation_rules_test.go:71
  in testGroupToString.func1
/Users/pingyu/workspace/go/src/github.com/pingcap/tidb/util/testutil/testutil.go:246
  in TestData.OnRecord
transformation_rules_test.go:69
  in testGroupToString
transformation_rules_test.go:157
  in testTransformationRuleSuite.TestProjectionElimination
/usr/local/Cellar/go/1.13/libexec/src/reflect/value.go:321
  in Value.Call
/Users/pingyu/workspace/go/pkg/mod/github.com/tiancaiamao/check@v0.0.0-20191119042138-8e73d07b629d/check.go:854
  in suiteRunner.forkTest.func1
/Users/pingyu/workspace/go/pkg/mod/github.com/tiancaiamao/check@v0.0.0-20191119042138-8e73d07b629d/check.go:739
  in suiteRunner.forkCall.func1
/usr/local/Cellar/go/1.13/libexec/src/runtime/asm_amd64.s:1357
  in goexit
```

Reply:
Emmm.. Sure, you can just keep it here now.